README.md
# BMI Calculator Package Documentation

## Introduction

This package, `BMI`, provides a way to compute the Body Mass Index (BMI) and interpret the results based on age, gender, weight, and height. The interpretations are based on guidelines from the World Health Organization (WHO).



### Properties:

- **height**: Height of the individual in meters.
- **weight**: Weight of the individual in kilograms.
- **ageYears**: Age of the individual in years.
- **ageMonths**: Additional age of the individual in months.
- **gender**: Gender of the individual. Accepts values "male" or "female".

### Methods:

1. **computeBMI()**
   - Computes the Body Mass Index (BMI) using the formula:
     BMI = kg/m2
   - Returns the computed BMI value as a double.

2. **interpretBMI()**
   - Interprets the BMI value:
     - For individuals outside the age range of 5-19 years, it uses the general BMI categories.
     - For children and teenagers aged 5-19, it uses BMI-for-age percentiles data, considering the gender of the individual.
   - Returns a string interpretation of the BMI.

## Interpretation Guidelines:

1. **For Adults (outside the age range of 5-19 years)**:
   - Severely underweight: BMI < 16
   - Underweight: 16 ≤ BMI < 16.9
   - Mildly underweight: 16.9 ≤ BMI < 18.4
   - Normal weight: 18.4 ≤ BMI < 24.9
   - Overweight: 24.9 ≤ BMI < 29.9
   - Obese Class I (Moderate): 29.9 ≤ BMI < 34.9
   - Obese Class II (Severe): 34.9 ≤ BMI < 39.9
   - Obese Class III (Very severe or morbidly obese): BMI ≥ 40

2. **For Children and Teenagers (age range of 5-19 years)**:
   - Severely wasted: BMI less than the -3rd percentile
   - Wasted: BMI between the -3rd and -2nd percentiles
   - Normal weight: BMI between the -2nd and 0th percentiles
   - Possible risk of overweight: BMI between the 0th and 1st percentiles
   - Overweight: BMI between the 1st and 2nd percentiles
   - Obese: BMI greater than the 2nd percentile

**Note**: For the age-specific interpretations, the percentile data from WHO is used. It's essential to ensure the correctness of this dataset for accurate interpretations.


CHANGELOG.md
## 1.0.0

* feat: added asian standard and changed classification ranges

## 0.0.3

* Shortened package description.

## 0.0.2

* Edited README.md

## 0.0.1

* Initial Release.




test/bmi_calculator_test.dart
import 'package:bmi_calculator/bmi_calculator.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('BMI class tests for males', () {
    test('Compute BMI', () {
      final bmiCalculator = BMI(
        standard: Standard.ASIAN,
        height: 1.75,
        weight: 70,
        ageYears: 25,
        ageMonths: 0,
        gender: "male",
      );

      expect(bmiCalculator.computeBMI(), closeTo(22.86, 0.01));
    });

    test('Interpret BMI for age not within 5-19', () {
      final bmiCalculator = BMI(
        standard: Standard.WHO,
        height: 1.75,
        weight: 70,
        ageYears: 25,
        ageMonths: 0,
        gender: "male",
      );

      expect(bmiCalculator.interpretBMI(), "Normal");
    });

    test('Interpret BMI for age within 5-19', () {
      final bmiCalculator = BMI(
        standard: Standard.ASIAN,
        height: 1.2,
        weight: 20,
        ageYears: 6,
        ageMonths: 5,
        gender: "male",
      );
      expect(bmiCalculator.interpretBMI(), "Normal weight");
    });
  });

  group('BMI class tests for females', () {
    test('Compute BMI for female', () {
      final bmiCalculator = BMI(
        standard: Standard.ASIAN,
        height: 1.6,
        weight: 55,
        ageYears: 28,
        ageMonths: 0,
        gender: "female",
      );

      expect(bmiCalculator.computeBMI(), closeTo(21.48, 0.01));
    });

    test('Interpret BMI for female age not within 5-19', () {
      final bmiCalculator = BMI(
        standard: Standard.WHO,
        height: 1.6,
        weight: 55,
        ageYears: 28,
        ageMonths: 0,
        gender: "female",
      );

      expect(bmiCalculator.interpretBMI(), "Normal");
    });

    test('Interpret BMI for female age within 5-19 (Normal Weight)', () {
      final bmiCalculator = BMI(
        standard: Standard.ASIAN,
        height: 1.4,
        weight: 30,
        ageYears: 10,
        ageMonths: 6,
        gender: "female",
      );
      expect(bmiCalculator.interpretBMI(), "Normal weight");
    });

    test('Interpret BMI for female age within 5-19 (Wasted)', () {
      final bmiCalculator = BMI(
        standard: Standard.ASIAN,
        height: 1.4,
        weight: 25,
        ageYears: 10,
        ageMonths: 6,
        gender: "female",
      );
      expect(bmiCalculator.interpretBMI(), "Wasted");
    });

    test('Interpret BMI for female age within 5-19 (Overweight)', () {
      final bmiCalculator = BMI(
        standard: Standard.ASIAN,
        height: 1.4,
        weight: 40,
        ageYears: 10,
        ageMonths: 6,
        gender: "female",
      );
      expect(bmiCalculator.interpretBMI(), "Overweight");
    });
  });
}


lib/bmi_calculator.dart
import 'package:bmi_calculator/data/boys_data.dart';
import 'package:bmi_calculator/data/girls_data.dart';

enum Standard { WHO, ASIAN }

class BMI {
  double height; //  meters
  double weight; //  kilograms
  int ageYears;
  int ageMonths;
  String gender;
  Standard standard;

  BMI({
    required this.height,
    required this.weight,
    required this.ageYears,
    required this.ageMonths,
    required this.gender,
    required this.standard,
  });

  double computeBMI() {
    return weight / (height * height);
  }

  String interpretBMI() {
    double bmi = computeBMI();

    if (ageYears < 5 || ageYears > 19) {
      // Compute BMI normally
      if(standard == Standard.WHO) {
        if (bmi < 18.5) {
          return "Underweight";
        } else if (bmi < 25) {
          return "Normal";
        } else if (bmi < 30) {
          return "Overweight";
        } else {
          return "Obese";
        }
      }
      else {
        if (bmi < 18.5) {
          return "Underweight";
        } else if (bmi < 23) {
          return "Normal";
        } else if (bmi < 25) {
          return "Overweight";
        } else {
          return "Obese";
        }
      }
    } else {
      // Use the gender-specific data
      Map<String, Map<String, double>> data =
          (gender == "male") ? boysBMI : girlsBMI;
      String ageKey = "$ageYears-$ageMonths";
      Map<String, double> ageData = data[ageKey] ?? {};

      // Find the appropriate percentile based on the data
      if (bmi < ageData['-3']!) {
        return "Severely wasted";
      } else if (bmi < ageData['-2']!) {
        return "Wasted";
      } else if (bmi < ageData['-1']!) {
        return "Normal weight";
      } else if (bmi < ageData['0']!) {
        return "Normal weight";
      } else if (bmi < ageData['1']!) {
        return "Possible risk of overweight";
      } else if (bmi < ageData['2']!) {
        return "Overweight";
      } else {
        return "Obese";
      }
    }
  }
}
